core/endorser/endorser.go

```diff
@@ -472,86 +473,124 @@ func (e *Endorser) ProcessProposalSuccessfullyOrError(up *UnpackedProposal) (*pb
 	// build a structure that collates all the information needed for the chaincode interest:
-	// 1. chaincodes in public RWset that don't have collections
-	// 2. collections in the private write set
-	// 3. collections in the PrivateReads struct
-	chaincodes := map[string]map[string]bool{} // map namespace -> collection -> isRead
-
-	// 1. Add any chaincodes (namespace) from the public RW set
-	if simResult.PubSimulationResults != nil {
-		for _, nsrws := range simResult.PubSimulationResults.GetNsRwset() {
-			if e.Support.IsSysCC(nsrws.Namespace) {
-				// skip system chaincodes
-				continue
-			}
-			chaincodes[nsrws.Namespace] = nil // no collections (maybe added later)
-		}
+	policies, err := parseWritesetMetadata(simResult.WritesetMetadata)
+	if err != nil {
+		return nil, err
 	}
 
-	// 2. Add any chaincodes/collections from the private write set
-	if simResult.PvtSimulationResults != nil {
-		// list collections that are read from (in order to populate NoPublicReads in ChaincodeInterest)
-		for _, prws := range simResult.PvtSimulationResults.GetNsPvtRwset() {
-			if e.Support.IsSysCC(prws.Namespace) {
-				// skip system chaincodes
-				continue
-			}
-			// build a map of collections that are read from or written to - value is true if read from, false if only written to
-			collections := map[string]bool{}
-			for _, cprws := range prws.GetCollectionPvtRwset() {
-				collections[cprws.CollectionName] = false
-			}
-			chaincodes[prws.Namespace] = collections
+	// There might be public states that are read and not written.  Need to add these to the policyRequired structure.
+	for _, nsrws := range simResult.PubSimulationResults.GetNsRwset() {
+		if e.Support.IsSysCC(nsrws.Namespace) {
+			// skip system chaincodes
+			continue
+		}
+		if _, ok := policies.policyRequired[nsrws.Namespace]; !ok {
+			// There's a public RWset for this namespace, but no public or private writes, so chaincode policy is required.
+			policies.add(nsrws.Namespace, "", true)
 		}
 	}
 
-	// 3. Add/merge collections from the PrivateReads struct
-	for ns, entry := range simResult.PrivateReads {
-		if e.Support.IsSysCC(ns) {
+	// There might be private collection states that are read and not written.  Need to add these to the policyRequired structure.
+	for chaincode, entry := range simResult.PrivateReads {
+		if e.Support.IsSysCC(chaincode) {
 			// skip system chaincodes
 			continue
 		}
-		if cc := chaincodes[ns]; cc == nil {
-			// chaincode either not already in the map, or nil entry - add it
-			chaincodes[ns] = map[string]bool{}
-		}
-		for name := range entry {
-			// even if this already exists from the write-set, it will override the value with true to indicate it was read
-			chaincodes[ns][name] = true
+		for collection := range entry {
+			if !policies.requireCollectionPolicy(chaincode, collection) {
+				policies.add(chaincode, "", true)
+			}
```

Comment:
Do you need this block at all? If there would be an entry in `PrivateReads` then the corresponding entry would be present in the hashed form in the public rwset which you have already looped through above.

Reply:
That's a good observation.  I had a spurious unit test that required this block, but I've fixed the test and removed this block.  The integration tests continue to pass which proves you are right :)